nuxeo-features/nuxeo-automation/nuxeo-automation-server/src/main/java/org/nuxeo/ecm/automation/server/jaxrs/adapters/AsyncOperationAdapter.java

```diff
@@ -0,0 +1,383 @@
+            }
+
+            // if output has a "url" key assume it's a redirect url
+            if (output instanceof Map) {
+                Object url = ((Map) output).get("url");
+                if (url instanceof String) {
+                    String baseUrl = VirtualHostHelper.getBaseURL(ctx.getRequest());
+                    return redirect(baseUrl + url);
+                }
+            }
+            return ResponseHelper.getResponse(output, request, HttpServletResponse.SC_OK);
+        }
+
+        throw new WebResourceNotFoundException("Execution with id=" + executionId + " not found");
+    }
+
+    @DELETE
+    @Path("{executionId}")
+    public Object abort(@PathParam("executionId") String executionId) throws IOException, MessagingException {
+        if (exists(executionId) && !isCompleted(executionId)) {
+            // TODO NXP-26304: support aborting any execution
+            if (isAsync(executionId)) {
+                Serializable taskId = getTaskId(executionId);
+                return getAsyncService(executionId).abort(taskId);
+            }
+            return ResponseHelper.getResponse("RUNNING", request, HttpServletResponse.SC_OK);
+        }
+        throw new WebResourceNotFoundException("Execution with id=" + executionId + " has completed");
+    }
+
+    protected TransientStore getTransientStore() {
+        return Framework.getService(TransientStoreService.class).getStore(STATUS_STORE_NAME);
+    }
+
+    protected void enterMethod(String executionId, InvokableMethod method) {
+        Map<String, Serializable> parameters = new HashMap<>();
+        // AsyncService.class is default => not async
+        if (!AsyncService.class.equals(method.getAsyncService())) {
+            parameters.put(TRANSIENT_STORE_SERVICE, method.getAsyncService().getName());
+        }
+        // reset parameters
+        getTransientStore().putParameters(executionId, parameters);
+    }
+
+    protected void setError(String executionId, String error) {
+        getTransientStore().putParameter(executionId, TRANSIENT_STORE_ERROR, error);
+        setCompleted(executionId);
+    }
+
+    public String getError(String executionId) {
+        return (String) getTransientStore().getParameter(executionId, TRANSIENT_STORE_ERROR);
+    }
+
+    protected void setOutput(String executionId, Serializable output) {
+        // store only taskId for async tasks
+        if (isAsync(executionId)) {
+            Serializable taskId = output instanceof AsyncStatus ? ((AsyncStatus) output).getId() : output;
+            getTransientStore().putParameter(executionId, TRANSIENT_STORE_TASK_ID, taskId);
+        } else {
+            if (output instanceof DocumentModel) {
+                output = detach((DocumentModel) output);
+            } else if (output instanceof DocumentModelList) {
+                output = new DocumentModelListImpl(
+                        ((DocumentModelList) output).stream().map(this::detach).collect(Collectors.toList()));
```

Comment:
It would be simpler to make `detach()` return `void`, and just modify the list members in-place:
`((DocumentModelList) output).forEach(this::detach)`